utils/types-manager.js

```diff
@@ -0,0 +1,195 @@
+      typesRange = `~${versionRange}`;
+    }
+
+    const bestVersion = maxSatisfying(versions, typesRange);
+
+    // If no compatible version found, trigger interactive prompt
+    if (!bestVersion) {
+      return await promptForTypesVersion(
+        typesPackageName,
+        versionRange,
+        versions,
+        true
+      );
+    }
+
+    return bestVersion;
+  } catch (error) {
+    console.warn(
+      `Failed to resolve @types version for ${typesPackageName}: ${error.message}`
+    );
+    return null;
+  }
+};
+
+const getTypesPackages /*: GetTypesPackages */ = async (
+  externals,
+  root,
+  roots,
+  promptForTypesVersion
+) => {
+  const typesDeps = [];
+
+  for (const {name, range} of externals) {
+    // Skip if already a @types package
+    if (name.startsWith('@types/')) continue;
+
+    // Convert scoped packages: @babel/core → @types/babel__core
+    const typesPackageName = name.startsWith('@')
+      ? `@types/${name.slice(1).replace('/', '__')}`
+      : `@types/${name}`;
+
+    try {
+      // Check if main package has bundled types
+      const hasBundledTypes = await checkBundledTypes(name, range, root);
+      if (hasBundledTypes) {
+        console.log(
+          `${name} has bundled types, removing separate @types package`
+        );
+        await removeTypesPackage(name, roots);
+        continue;
+      }
+
+      // Find best @types version
+      const typesVersion = await findBestTypesVersion(
+        typesPackageName,
+        range,
+        root,
+        promptForTypesVersion
+      );
+      if (typesVersion) {
+        console.log(`Adding ${typesPackageName}@${typesVersion}`);
+        typesDeps.push(`${typesPackageName}@${typesVersion}`);
+      }
+    } catch (error) {
+      console.warn(`Failed to process types for ${name}: ${error.message}`);
+      // Continue processing other packages
+    }
+  }
+
+  return typesDeps;
+};
+
+const removeTypesPackage /*: RemoveTypesPackage */ = async (
+  packageName,
+  roots
+) => {
+  const {read, write} = require('./node-helpers.js');
+  // Convert scoped packages: @babel/core → @types/babel__core
+  const typesPackageName = packageName.startsWith('@')
+    ? `@types/${packageName.slice(1).replace('/', '__')}`
+    : `@types/${packageName}`;
+
+  await Promise.all(
+    roots.map(async cwd => {
+      const meta = JSON.parse(await read(`${cwd}/package.json`, 'utf8'));
+      let updated = false;
+
+      // Remove from dependencies and devDependencies
+      if (meta.dependencies && meta.dependencies[typesPackageName]) {
+        delete meta.dependencies[typesPackageName];
+        updated = true;
+      }
+      if (meta.devDependencies && meta.devDependencies[typesPackageName]) {
+        delete meta.devDependencies[typesPackageName];
+        updated = true;
+      }
+
+      if (updated) {
+        await write(
+          `${cwd}/package.json`,
+          JSON.stringify(meta, null, 2) + '\n',
```

Comment:
Also feels like it could be encapsulated in some utility helper, given we use a similar format elsewhere in this project.  May not be worth updating, but good to keep in mind.